src/aspire/reconstruction/mean.py

```diff
@@ -1,17 +1,262 @@
+            for i in range(0, self.src.n, self.batch_size):
+                im = self.src.images[i : i + self.batch_size]
+
+                batch_mean_b = (
+                    self.src.im_backward(im, i, self.weights[:, k]) / self.src.n
+                )
+                mean_b[k] += batch_mean_b.astype(self.dtype)
+
+        res = np.sqrt(self.src.n) * self.basis.evaluate_t(mean_b)
+        logger.info(f"Determined weighted adjoint mappings. Shape = {res.shape}")
+
+        return res
+
+    def conj_grad(self, b_coeff, tol=1e-5, regularizer=0):
+        n = b_coeff.shape[-1]  # 0???
+        kernel = self.kernel
+
+        if regularizer > 0:
+            kernel += regularizer
+
+        operator = LinearOperator(
+            (self.r * n, self.r * n),
+            matvec=partial(self.apply_kernel, kernel=kernel),
+            dtype=self.dtype,
+        )
+        if self.precond_kernel is None:
+            M = None
+        else:
+            precond_kernel = self.precond_kernel
+            if regularizer > 0:
+                precond_kernel += regularizer
+            M = LinearOperator(
+                (self.r * n, self.r * n),
+                matvec=partial(self.apply_kernel, kernel=precond_kernel),
+                dtype=self.dtype,
+            )
+
+        tol = tol or config.mean.cg_tol
+        target_residual = tol * norm(b_coeff)
+
+        def cb(xk):
+            logger.info(
+                f"Delta {norm(b_coeff - self.apply_kernel(xk))} (target {target_residual})"
+            )
+
+        x, info = cg(operator, b_coeff.flatten(), M=M, callback=cb, tol=tol, atol=0)
+
+        if info != 0:
+            raise RuntimeError("Unable to converge!")
+
+        # Thinking might be clearer if r x ... but would need to mess with roll/unroll in FBB.
+        # return x.reshape(self.r, -1)
+        return x.reshape(self.r, self.basis.count)
+
+    def apply_kernel(self, vol_coeff, kernel=None):
+        """
+        Applies the kernel represented by convolution
+
+        :param vol_coeff: The volume to be convolved, stored in the basis coefficients.
+        :param kernel: a Kernel object. If None, the kernel for this Estimator is used.
+        :return: The result of evaluating `vol_coeff` in the given basis, convolving with the kernel given by
+            kernel, and backprojecting into the basis.
+        """
+        if kernel is None:
+            kernel = self.kernel
+
+        assert np.size(vol_coeff) == self.r * self.basis.count
+        if vol_coeff.ndim == 1:
+            vol_coeff = vol_coeff.reshape(self.r, self.basis.count)
+
+        vols_out = Volume(
+            np.zeros((self.r, self.src.L, self.src.L, self.src.L), dtype=self.dtype)
+        )
+
+        vol = self.basis.evaluate(vol_coeff)
+
+        for k in range(self.r):
+            for j in range(self.r):
+                vols_out[k] = vols_out[k] + kernel.convolve_volume(vol[j], j, k)
+                # Note this is where we would add mask_gamma
```

Comment:
?

Reply:
If remembering correctly, `mask_gamma` was a feature Amit was working on that caused the version of the code I was given to be different from the better version in the `het3d` repo.  I don't know if that work was completed. This was a note to myself in case I needed to find the spot quickly. Can remove the comment.

Reply:
This is something we can revisit it needed. Might as well leave the comment in.